# Dashboard Demo

Generate the dataset, create interactive charts with filters, and build the final dashboard in `dashboard.html`.

In [1]:
import pandas as pd
import plotly.graph_objects as go
import numpy as np
import json
from collections import Counter

## 1. Sample Dataset

# ============================================
# CONFIGURE HERE: replace these values with your own data

In [2]:
# ============================================
months = ['January', 'February', 'March', 'April', 'May', 'June']
monthly_data = pd.DataFrame({
    'Mes': months,
    'Sales': [45000, 52000, 48000, 61000, 58000, 67000],
    'Units': [180, 210, 195, 240, 225, 265]
})

products = ['Product A', 'Product B', 'Product C']
product_data = pd.DataFrame({
    'Product': products,
    'Revenue': [48000, 31500, 63000],
    'Units': [320, 210, 150]
})

product_data

,Product,Revenue,Units
0,Product A,48000,320
1,Product B,31500,210
2,Product C,63000,150


## 2. Chart 1 — Line Chart with Filter (Sales / Units)

This chart goes on the **left side**, spanning the full height.

In [3]:
line_chart = go.Figure()
line_chart.add_trace(go.Scatter(x=monthly_data['Mes'], y=monthly_data['Sales'],
                                mode='lines+markers', name='Sales',
                                line=dict(color='#2563eb', width=3)))
line_chart.add_trace(go.Scatter(x=monthly_data['Mes'], y=monthly_data['Units'],
                                mode='lines+markers', name='Units',
                                line=dict(color='#16a34a', width=3), visible=False))

line_chart.update_layout(
    title='Monthly Trend',
    updatemenus=[dict(
        type='dropdown', direction='down', x=1, xanchor='right', y=1.15,
        buttons=[
            dict(label='Ver Sales', method='update',
                 args=[{'visible': [True, False]}, {'title': 'Monthly Trend — Sales'}]),
            dict(label='Ver Units', method='update',
                 args=[{'visible': [False, True]}, {'title': 'Monthly Trend — Units'}]),
        ]
    )]
)
line_chart.show()

## 3. Chart 2 — Bar Chart with Filter (Revenue / Units)

This chart goes in the **top-right** section.

In [4]:
bar_chart = go.Figure()
bar_chart.add_trace(go.Bar(x=product_data['Product'], y=product_data['Revenue'],
                             name='Revenue', marker_color='#2563eb', visible=True))
bar_chart.add_trace(go.Bar(x=product_data['Product'], y=product_data['Units'],
                             name='Units', marker_color='#f59e0b', visible=False))

bar_chart.update_layout(
    title='Por Product — Revenue',
    updatemenus=[dict(
        type='dropdown', direction='down', x=1, xanchor='right', y=1.2,
        buttons=[
            dict(label='Ver Revenue', method='update',
                 args=[{'visible': [True, False]}, {'title': 'Por Product — Revenue'}]),
            dict(label='Ver Units', method='update',
                 args=[{'visible': [False, True]}, {'title': 'Por Product — Units'}]),
        ]
    )]
)
bar_chart.show()

## 4. Chart 3 — Pie Chart (Unit Distribution)

This chart goes in the **bottom-right** section. Its filtering is native: click the legend to hide or show each product.

In [5]:
pie_chart = go.Figure(data=[go.Pie(labels=product_data['Product'], values=product_data['Units'])])
pie_chart.update_layout(title='Distribución de Units')
pie_chart.show()

## 5. Build the Final Dashboard (`dashboard.html`)

This section defines the **order and layout** of the charts in the grid. To rearrange them, change which figure is assigned to each position (`left_chart`, `top_right_chart`, `bottom_right_chart`).

In [6]:
# ============================================
# CONFIGURE HERE: assign each chart to a grid position
# ============================================
left_chart = line_chart
top_right_chart = bar_chart
bottom_right_chart = pie_chart

left_html = left_chart.to_html(full_html=False, include_plotlyjs='cdn', config={'responsive': True})
top_html = top_right_chart.to_html(full_html=False, include_plotlyjs=False, config={'responsive': True})
bottom_html = bottom_right_chart.to_html(full_html=False, include_plotlyjs=False, config={'responsive': True})

template = '''<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<title>Dashboard</title>
<style>
  * { box-sizing: border-box; margin: 0; padding: 0; font-family: 'Segoe UI', Arial, sans-serif; }
  body { background: #f3f4f6; padding: 20px; }
  .dashboard-grid {
    display: grid;
    grid-template-columns: 1.3fr 1fr;
    grid-template-rows: 1fr 1fr;
    gap: 16px;
    height: calc(100vh - 40px);
  }
  .panel {
    background: white;
    border-radius: 12px;
    box-shadow: 0 2px 10px rgba(0,0,0,0.06);
    padding: 12px;
    overflow: hidden;
  }
  .panel-left { grid-column: 1; grid-row: 1 / 3; }
  .panel-top-right { grid-column: 2; grid-row: 1; }
  .panel-bottom-right { grid-column: 2; grid-row: 2; }
  .panel > div { width: 100% !important; height: 100% !important; }
</style>
</head>
<body>
  <div class="dashboard-grid">
    <div class="panel panel-left">__LEFT__</div>
    <div class="panel panel-top-right">__TOP_RIGHT__</div>
    <div class="panel panel-bottom-right">__BOTTOM_RIGHT__</div>
  </div>
</body>
</html>
'''

template = template.replace('__LEFT__', left_html)
template = template.replace('__TOP_RIGHT__', top_html)
template = template.replace('__BOTTOM_RIGHT__', bottom_html)

with open('dashboard.html', 'w', encoding='utf-8') as f:
    f.write(template)

print("dashboard.html generated")

dashboard.html generated


## 6. Data Quality — Load the Dataset

Load the real dataset used by the Data Quality dashboard in `data_quality.html`.

In [7]:
# ============================================
# CONFIGURE HERE: GitHub repo info
# ============================================
github_user = "elvaramosglz"
github_repo = "dashboard-demo"
github_branch = "main"
csv_filename = "dataset.csv"
dataset_encoding = "mac_roman"  # this file is saved as mac_roman, not UTF-8 (e.g. "Maracan\u00e3" breaks under utf-8)

# Built automatically \u2014 don't need to edit this line
dataset_url = f"https://raw.githubusercontent.com/{github_user}/{github_repo}/{github_branch}/{csv_filename}"

# Reads straight from GitHub, so anyone who opens this notebook gets the same
# CSV you have in the repo \u2014 no need to have the file saved locally
df_quality = pd.read_csv(dataset_url, encoding=dataset_encoding)
df_quality.head()

,MONTH,CITY,TRIPS,GROSS BOOKINGS (GB),SHOPPING SESSIONS,RIDERS WITH SESSIONS,REQUESTING SESSIONS,REQUESTS,ACTIVE DRIVERS,SUPPLY HOURS,...,RS/SS,C/SS,ACCEPTANCE RATE,DRIVER CANCELATION,SUPPLY UTILIZATION,NPI,DRIVER PAYMENTS (AS % OF GB),RIDER INCENTIVES AND PROMOTIONS (AS % OF GB),DRIVER INCENTIVES AND PROMOTIONS (AS % OF GB),TAKE RATE (AS % OF GB)
0,2022-01,Vila Belmiro,"3,005,706","56,529,569","5,283,006","711,432","3,257,826","3,844,423","18,795","1,564,816",...,61.7%,56.9%,23.0%,20.0%,63.0%,1.03,73.5%,5.7%,5.1%,14.2%
1,2022-01,Rose Bowl,"1,533,078","25,315,829","2,692,591","406,844","1,638,051","1,859,779","8,742","824,750",...,60.8%,56.9%,34.0%,15.0%,59.0%,0.98,76.3%,-3.3%,4.3%,20.5%
2,2022-01,San Siro,"8,108,570","155,667,588","14,385,079","1,609,850","8,570,326","9,515,128","50,056","4,187,961",...,59.6%,56.4%,44.0%,22.0%,73.0%,1.02,78.8%,4.9%,3.6%,10.6%
3,2022-01,Lusail,"2,163,550","38,225,180","3,939,724","493,960","2,348,175","2,859,544","10,845","1,074,968",...,59.6%,54.9%,25.0%,20.0%,70.0%,1.05,76.8%,6.2%,3.6%,11.8%
4,2022-01,Old Trafford,"11,042,462","206,436,144","19,113,887","2,553,885","11,801,188","13,152,423","63,551","5,202,255",...,61.7%,57.8%,46.0%,27.0%,77.0%,1.05,75.6%,-1.4%,3.4%,17.0%


## 7. Data Quality — Completeness Calculations

Calculate completion percentage, null count, and city/month coverage directly from the dataset.

In [8]:
total_rows = len(df_quality)
total_cols = len(df_quality.columns)
total_cells = total_rows * total_cols

null_count = int(df_quality.isna().sum().sum())
completeness_pct = round((1 - null_count / total_cells) * 100, 1)
completeness_pct_label = f"{completeness_pct:.0f}" if completeness_pct == int(completeness_pct) else f"{completeness_pct}"

n_cities = df_quality['CITY'].nunique()
n_months = df_quality['MONTH'].nunique()

if null_count == 0:
    summary_text = f"{total_rows} records \u00b7 {n_cities} cities \u00b7 {n_months} months \u00b7 no missing values"
else:
    summary_text = f"{total_rows} records \u00b7 {n_cities} cities \u00b7 {n_months} months \u00b7 {null_count} missing values"

completeness_pct, null_count, summary_text

(100.0, 0, '168 records · 8 cities · 21 months · no missing values')

## 8. Build the Data Quality Card (`data_quality.html`)

Generate a compact, centered Uber-blue hero card using the values calculated above.

In [9]:
# ============================================
# CONFIGURE HERE: colors and label
# ============================================
uber_blue = '#276EF1'
card_label = 'Data Completeness'

data_quality_html = f'''<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<title>Data Quality</title>
<style>
  * {{ box-sizing: border-box; margin: 0; padding: 0; font-family: 'Segoe UI', Arial, sans-serif; }}
  body {{
    background: #f3f4f6;
    padding: 24px;
    min-height: 100vh;
    display: flex;
    align-items: center;
    justify-content: center;
  }}

  .hero-card {{
    background: {uber_blue};
    color: #FFFFFF;
    border-radius: 16px;
    padding: 28px 40px;
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    gap: 4px;
    box-shadow: 0 4px 20px rgba(39,110,241,0.25);
    max-width: 320px;
  }}
  .hero-card .check {{ font-size: 32px; line-height: 1; margin-bottom: 4px; }}
  .hero-card .hero-number {{ font-size: 44px; font-weight: 700; line-height: 1; }}
  .hero-card .hero-label {{ font-size: 14px; font-weight: 600; opacity: 0.95; }}
  .hero-card .hero-sub {{ font-size: 11px; opacity: 0.8; }}
</style>
</head>
<body>

  <div class="hero-card">
    <div class="check">\u2705</div>
    <div class="hero-number">{completeness_pct_label}%</div>
    <div class="hero-label">{card_label}</div>
    <div class="hero-sub">{summary_text}</div>
  </div>

</body>
</html>
'''

with open('data_quality.html', 'w', encoding='utf-8') as f:
    f.write(data_quality_html)

print("data_quality.html generated")

data_quality.html generated


## 9. Data Quality — Per-Column Breakdown

Calculate completeness, null count, and status for each column. These values populate the dropdown in `data_quality.html`.

In [10]:
# ============================================
# CONFIGURE HERE: short description per column
# and the OK / warning / critical thresholds
# NOTE: "NPI" description below is a placeholder — confirm the real definition
# ============================================
column_descriptions = {
    "MONTH": "Reporting month for the record (YYYY-MM).",
    "CITY": "Market / city where the metrics were recorded.",
    "TRIPS": "Number of completed trips.",
    "GROSS BOOKINGS (GB)": "Total gross bookings value generated.",
    "SHOPPING SESSIONS": "App sessions where a rider browsed, not necessarily requesting a ride.",
    "RIDERS WITH SESSIONS": "Unique riders who opened a session.",
    "REQUESTING SESSIONS": "Sessions where the rider went on to request a ride.",
    "REQUESTS": "Total ride requests made.",
    "ACTIVE DRIVERS": "Unique drivers who were online during the period.",
    "SUPPLY HOURS": "Total hours drivers spent online and available.",
    "C/R": "Completion rate: completed trips over requests.",
    "C/RS": "Completion rate: completed trips over requesting sessions.",
    "RS/SS": "Conversion rate: requesting sessions over shopping sessions.",
    "C/SS": "Completion rate: completed trips over shopping sessions.",
    "ACCEPTANCE RATE": "Share of requests accepted by drivers.",
    "DRIVER CANCELATION": "Share of trips cancelled by drivers.",
    "SUPPLY UTILIZATION": "Share of available driver supply hours actually used.",
    "NPI": "Placeholder \u2014 confirm real definition.",
    "DRIVER PAYMENTS (AS % OF GB)": "Driver earnings as a percentage of gross bookings.",
    "RIDER INCENTIVES AND PROMOTIONS (AS % OF GB)": "Rider-side incentives/promos as a percentage of gross bookings.",
    "DRIVER INCENTIVES AND PROMOTIONS (AS % OF GB)": "Driver-side incentives/promos as a percentage of gross bookings.",
    "TAKE RATE (AS % OF GB)": "Platform\'s take rate as a percentage of gross bookings.",
}

ok_threshold = 98      # >= this % -> OK
warn_threshold = 90    # >= this % (and below ok_threshold) -> warning

column_stats = []
for col in df_quality.columns:
    nulls = int(df_quality[col].isna().sum())
    pct = round((1 - nulls / total_rows) * 100, 1)
    pct_label = f"{pct:.0f}" if pct == int(pct) else f"{pct}"
    if pct >= ok_threshold:
        status = "ok"
    elif pct >= warn_threshold:
        status = "warn"
    else:
        status = "bad"
    column_stats.append({
        "name": col,
        "description": column_descriptions.get(col, "No description available."),
        "completeness": pct_label,
        "nulls": nulls,
        "total": total_rows,
        "status": status,
    })

column_stats

[{'name': 'MONTH',
  'description': 'Reporting month for the record (YYYY-MM).',
  'completeness': '100',
  'nulls': 0,
  'total': 168,
  'status': 'ok'},
 {'name': 'CITY',
  'description': 'Market / city where the metrics were recorded.',
  'completeness': '100',
  'nulls': 0,
  'total': 168,
  'status': 'ok'},
 {'name': 'TRIPS',
  'description': 'Number of completed trips.',
  'completeness': '100',
  'nulls': 0,
  'total': 168,
  'status': 'ok'},
 {'name': 'GROSS BOOKINGS (GB)',
  'description': 'Total gross bookings value generated.',
  'completeness': '100',
  'nulls': 0,
  'total': 168,
  'status': 'ok'},
 {'name': 'SHOPPING SESSIONS',
  'description': 'App sessions where a rider browsed, not necessarily requesting a ride.',
  'completeness': '100',
  'nulls': 0,
  'total': 168,
  'status': 'ok'},
 {'name': 'RIDERS WITH SESSIONS',
  'description': 'Unique riders who opened a session.',
  'completeness': '100',
  'nulls': 0,
  'total': 168,
  'status': 'ok'},
 {'name': 'REQUESTING

## 10. Build the Data Nulls Section (`data_quality.html`)

Title, description, KPI card, and a dropdown to drill into each column \u2014 all populated from `column_stats` above, so editing the CSV updates this automatically.

In [11]:
# ============================================
# CONFIGURE HERE: titles, copy and colors
# ============================================
uber_blue = "#276EF1"
uber_green = "#3AA76D"
uber_yellow = "#FFC043"
uber_red = "#D44333"

page_title = "DATA NULLS"
page_description = (
    "This check scans every column in the dataset for missing values, "
    "so you can see at a glance which fields are complete and which ones "
    "need attention. Pick a column below to see its details."
)

status_colors = {"ok": uber_green, "warn": uber_yellow, "bad": uber_red}
status_labels = {"ok": "OK", "warn": "Review", "bad": "Critical"}

column_stats_json = json.dumps(column_stats)

data_quality_html = f'''<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<title>Data Quality</title>
<style>
  * {{ box-sizing: border-box; margin: 0; padding: 0; font-family: 'Segoe UI', Arial, sans-serif; }}
  body {{ background: #f3f4f6; padding: 32px; }}
  .wrap {{ max-width: 640px; margin: 0 auto; }}
  h2 {{ color: #111; margin-bottom: 8px; text-align: center; text-transform: uppercase; letter-spacing: 0.5px; }}
  .description {{ color: #6b7280; font-size: 14px; margin-bottom: 24px; line-height: 1.5; text-align: center; }}

  .hero-card {{
    background: {uber_blue};
    color: #FFFFFF;
    border-radius: 16px;
    padding: 20px 32px;
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    gap: 4px;
    margin: 0 auto 24px auto;
    max-width: 280px;
  }}
  .hero-card .hero-number {{ font-size: 40px; font-weight: 700; line-height: 1; }}
  .hero-card .hero-label {{ font-size: 13px; font-weight: 600; opacity: 0.95; }}
  .hero-card .hero-sub {{ font-size: 11px; opacity: 0.8; }}

  .section-card {{ background: white; border-radius: 12px; padding: 28px; box-shadow: 0 2px 10px rgba(0,0,0,0.06); }}
  .section-card label {{ font-size: 13px; color: #6b7280; display: block; margin-bottom: 8px; }}
  select {{
    width: 100%;
    padding: 10px 12px;
    border-radius: 8px;
    border: 1px solid #d1d5db;
    font-size: 14px;
    margin-bottom: 20px;
    background: white;
  }}

  .col-name {{ font-size: 18px; font-weight: 700; color: #111; margin-bottom: 10px; }}
  .metric-row {{ font-size: 14px; color: #374151; margin-bottom: 6px; }}
  .metric-label {{ color: #6b7280; }}
  #col-pct {{ font-weight: 700; }}
  .badge {{ padding: 3px 10px; border-radius: 999px; font-size: 12px; font-weight: 600; }}
</style>
</head>
<body>
  <div class="wrap">
    <div class="section-card">
      <h2>{page_title}</h2>
      <p class="description">{page_description}</p>

      <div class="hero-card">
        <div class="hero-number">{completeness_pct_label}%</div>
        <div class="hero-label">Data Completeness</div>
        <div class="hero-sub">{summary_text}</div>
      </div>

      <label for="column-select">Choose a column</label>
      <select id="column-select" onchange="renderColumn(this.value)"></select>

      <div class="col-name" id="col-name"></div>
      <div class="metric-row"><span class="metric-label">Completion rate:</span> <span id="col-pct"></span> <span class="badge" id="col-badge"></span></div>
      <div class="metric-row"><span class="metric-label">Total nulls:</span> <span id="col-nulls"></span></div>
      <div class="metric-row"><span class="metric-label">Total data:</span> <span id="col-total"></span></div>
    </div>
  </div>

<script>
  const columnStats = {column_stats_json};
  const statusColors = {{ ok: "{uber_green}", warn: "{uber_yellow}", bad: "{uber_red}" }};
  const statusLabels = {{ ok: "OK", warn: "Review", bad: "Critical" }};

  const select = document.getElementById('column-select');
  columnStats.forEach(function(col) {{
    const opt = document.createElement('option');
    opt.value = col.name;
    opt.textContent = col.name;
    select.appendChild(opt);
  }});

  function renderColumn(name) {{
    const col = columnStats.find(function(c) {{ return c.name === name; }});
    if (!col) return;
    document.getElementById('col-name').textContent = col.name;
    document.getElementById('col-pct').textContent = col.completeness + "%";
    document.getElementById('col-pct').style.color = statusColors[col.status];
    document.getElementById('col-nulls').textContent = col.nulls;
    document.getElementById('col-total').textContent = col.total;
    const badge = document.getElementById('col-badge');
    badge.textContent = statusLabels[col.status];
    badge.style.background = statusColors[col.status] + "22";
    badge.style.color = statusColors[col.status];
  }}

  renderColumn(columnStats[0].name);
</script>
</body>
</html>
'''

with open("data_quality.html", "w", encoding="utf-8") as f:
    f.write(data_quality_html)

print("data_quality.html generated with", len(column_stats), "columns")

data_quality.html generated with 22 columns


## 11. Data Quality — Month Coverage by City

For each city, calculate how many months it *should* contain based on the full dataset date range, how many it actually contains, and any missing or duplicated months. These values populate the Month Coverage table in `data_quality.html`.

In [12]:
from collections import Counter

all_months = sorted(df_quality['MONTH'].unique())
expected_months = len(all_months)
first_month = all_months[0]
last_month = all_months[-1]

city_coverage = []
for city, g in df_quality.groupby('CITY'):
    months_list = g['MONTH'].tolist()
    unique_months = sorted(set(months_list))
    missing = sorted(set(all_months) - set(unique_months))
    dupes = sorted([m for m, c in Counter(months_list).items() if c > 1])
    coverage_pct = round(len(unique_months) / expected_months * 100, 1)
    coverage_label = f"{coverage_pct:.0f}" if coverage_pct == int(coverage_pct) else f"{coverage_pct}"

    if coverage_pct >= ok_threshold:
        cov_status = "ok"
    elif coverage_pct >= warn_threshold:
        cov_status = "warn"
    else:
        cov_status = "bad"

    city_coverage.append({
        "city": city,
        "expected_months": expected_months,
        "rows_found": len(g),
        "unique_months": len(unique_months),
        "coverage": coverage_label,
        "missing_months": ", ".join(missing) if missing else "None",
        "duplicate_months": ", ".join(dupes) if dupes else "None",
        "status": cov_status,
    })

city_coverage

[{'city': 'Azteca',
  'expected_months': 21,
  'rows_found': 21,
  'unique_months': 21,
  'coverage': '100',
  'missing_months': 'None',
  'duplicate_months': 'None',
  'status': 'ok'},
 {'city': 'Lusail',
  'expected_months': 21,
  'rows_found': 21,
  'unique_months': 21,
  'coverage': '100',
  'missing_months': 'None',
  'duplicate_months': 'None',
  'status': 'ok'},
 {'city': 'Maracanã',
  'expected_months': 21,
  'rows_found': 21,
  'unique_months': 21,
  'coverage': '100',
  'missing_months': 'None',
  'duplicate_months': 'None',
  'status': 'ok'},
 {'city': 'Old Trafford',
  'expected_months': 21,
  'rows_found': 21,
  'unique_months': 21,
  'coverage': '100',
  'missing_months': 'None',
  'duplicate_months': 'None',
  'status': 'ok'},
 {'city': 'Rose Bowl',
  'expected_months': 21,
  'rows_found': 21,
  'unique_months': 21,
  'coverage': '100',
  'missing_months': 'None',
  'duplicate_months': 'None',
  'status': 'ok'},
 {'city': 'San Siro',
  'expected_months': 21,
  'rows_foun

## 12. Add the Month Coverage Table to `data_quality.html`

Append a second section card below the existing section using `city_coverage`, so it updates automatically when the CSV changes.

In [13]:
# ============================================
# CONFIGURE HERE: table title and copy
# ============================================
coverage_title = "MONTH COVERAGE"
coverage_description = (
    f"Checks that every city has a row for each month between {first_month} and {last_month} "
    "(the full date range in the dataset), and flags any missing or duplicated months."
)

city_coverage_json = json.dumps(city_coverage)

coverage_rows_html = ""
for row in city_coverage:
    color = {"ok": uber_green, "warn": uber_yellow, "bad": uber_red}[row["status"]]
    badge_bg = color + "1A"
    coverage_rows_html += f'''
        <tr>
          <td class="city-cell">{row["city"]}</td>
          <td class="num-cell">{row["expected_months"]}</td>
          <td class="num-cell">{row["rows_found"]}</td>
          <td class="num-cell">{row["unique_months"]}</td>
          <td class="num-cell"><span class="cov-badge" style="color:{color}; background:{badge_bg};">{row["coverage"]}%</span></td>
          <td class="muted-cell">{row["missing_months"]}</td>
          <td class="muted-cell">{row["duplicate_months"]}</td>
        </tr>'''

coverage_section_html = f'''
    <div class="section-card" style="margin-top:24px;">
      <h2>{coverage_title}</h2>
      <p class="description">{coverage_description}</p>
      <div class="table-scroll">
      <table class="coverage-table">
        <thead>
          <tr>
            <th>City</th>
            <th>Expected Months</th>
            <th>Rows Found</th>
            <th>Unique Months</th>
            <th>Coverage %</th>
            <th>Missing Months</th>
            <th>Duplicate Months</th>
          </tr>
        </thead>
        <tbody>{coverage_rows_html}
        </tbody>
      </table>
      </div>
    </div>'''

# Insert the new section right before the closing </div> of .wrap, and add its CSS
# right before </style>, without touching anything already built above.
coverage_css = '''
  .wrap { max-width: 820px; }
  .table-scroll { overflow-x: auto; border-radius: 10px; border: 1px solid #f0f0f0; }
  .coverage-table { width: 100%; border-collapse: collapse; font-size: 13px; }
  .coverage-table th, .coverage-table td { padding: 12px 14px; white-space: nowrap; }
  .coverage-table thead th {
    text-align: center;
    color: #6b7280;
    font-weight: 700;
    font-size: 11px;
    text-transform: uppercase;
    letter-spacing: 0.4px;
    background: #f9fafb;
    border-bottom: 2px solid #eef0f3;
  }
  .coverage-table thead th:first-child { text-align: left; }
  .coverage-table tbody tr { border-bottom: 1px solid #f2f3f5; }
  .coverage-table tbody tr:last-child { border-bottom: none; }
  .coverage-table tbody tr:nth-child(even) { background: #fafbfc; }
  .coverage-table tbody tr:hover { background: #f2f6fe; }
  .city-cell { font-weight: 600; color: #111; text-align: left; }
  .num-cell { text-align: center; color: #374151; }
  .muted-cell { text-align: center; color: #9ca3af; font-size: 12px; }
  .cov-badge { display: inline-block; padding: 4px 12px; border-radius: 999px; font-weight: 700; }
'''

data_quality_html = data_quality_html.replace("</style>", coverage_css + "</style>")
data_quality_html = data_quality_html.replace("  </div>\n\n<script>", "  </div>\n" + coverage_section_html + "\n\n<script>")

with open("data_quality.html", "w", encoding="utf-8") as f:
    f.write(data_quality_html)

print("Month Coverage section added for", len(city_coverage), "cities")

Month Coverage section added for 8 cities


## 13. Data Quality — how pandas actually reads each column

This section reports the real pandas dtype, a representative Python scalar type, an example value, null count, and unique values. It does **not** use Excel's visual formats such as General, Number, or Percentage.

In [14]:
def python_scalar_type(series):
    """Returns the Python/NumPy type name of the first non-null value."""
    non_null = series.dropna()
    if non_null.empty:
        return "No non-null values"
    return type(non_null.iloc[0]).__name__


def dtype_family(series):
    """Groups pandas dtypes into business-friendly labels."""
    if pd.api.types.is_datetime64_any_dtype(series):
        return "datetime"
    if pd.api.types.is_bool_dtype(series):
        return "boolean"
    if pd.api.types.is_integer_dtype(series):
        return "integer"
    if pd.api.types.is_float_dtype(series):
        return "float"
    if pd.api.types.is_numeric_dtype(series):
        return "numeric"
    if pd.api.types.is_string_dtype(series) or series.dtype == "object":
        return "string/object"
    return str(series.dtype)


dtype_report = []
for col in df_quality.columns:
    non_null = df_quality[col].dropna()
    dtype_report.append({
        "COLUMN": col,
        "PANDAS_DTYPE": str(df_quality[col].dtype),
        "DTYPE_FAMILY": dtype_family(df_quality[col]),
        "PYTHON_TYPE_EXAMPLE": python_scalar_type(df_quality[col]),
        "EXAMPLE_VALUE": non_null.iloc[0] if not non_null.empty else None,
        "NULLS": int(df_quality[col].isna().sum()),
        "UNIQUE_VALUES": int(df_quality[col].nunique(dropna=True)),
    })

dtype_report = pd.DataFrame(dtype_report)
dtype_report

,COLUMN,PANDAS_DTYPE,DTYPE_FAMILY,PYTHON_TYPE_EXAMPLE,EXAMPLE_VALUE,NULLS,UNIQUE_VALUES
0,MONTH,str,string/object,str,2022-01,0,21
1,CITY,str,string/object,str,Vila Belmiro,0,8
2,TRIPS,str,string/object,str,"3,005,706",0,168
3,GROSS BOOKINGS (GB),str,string/object,str,"56,529,569",0,168
4,SHOPPING SESSIONS,str,string/object,str,"5,283,006",0,168
5,RIDERS WITH SESSIONS,str,string/object,str,"711,432",0,168
6,REQUESTING SESSIONS,str,string/object,str,"3,257,826",0,168
7,REQUESTS,str,string/object,str,"3,844,423",0,168
8,ACTIVE DRIVERS,str,string/object,str,"18,795",0,168
9,SUPPLY HOURS,str,string/object,str,"1,564,816",0,168


## 14. Data Quality — expected analytical type vs. actual pandas type

The expected type is based on what each metric represents. Percentage columns are expected to become numeric ratios for analysis, even when the raw CSV contains `%` signs.

In [15]:
expected_types = {
    "MONTH": "datetime",
    "CITY": "string",
    "TRIPS": "integer",
    "GROSS BOOKINGS (GB)": "float",
    "SHOPPING SESSIONS": "integer",
    "RIDERS WITH SESSIONS": "integer",
    "REQUESTING SESSIONS": "integer",
    "REQUESTS": "integer",
    "ACTIVE DRIVERS": "integer",
    "SUPPLY HOURS": "float",
    "C/R": "float",
    "C/RS": "float",
    "RS/SS": "float",
    "C/SS": "float",
    "ACCEPTANCE RATE": "float",
    "DRIVER CANCELATION": "float",
    "SUPPLY UTILIZATION": "float",
    "NPI": "float",
    "DRIVER PAYMENTS (AS % OF GB)": "float",
    "RIDER INCENTIVES AND PROMOTIONS (AS % OF GB)": "float",
    "DRIVER INCENTIVES AND PROMOTIONS (AS % OF GB)": "float",
    "TAKE RATE (AS % OF GB)": "float",
}

percentage_columns = {
    "C/R", "C/RS", "RS/SS", "C/SS", "ACCEPTANCE RATE",
    "DRIVER CANCELATION", "SUPPLY UTILIZATION",
    "DRIVER PAYMENTS (AS % OF GB)",
    "RIDER INCENTIVES AND PROMOTIONS (AS % OF GB)",
    "DRIVER INCENTIVES AND PROMOTIONS (AS % OF GB)",
    "TAKE RATE (AS % OF GB)",
}

integer_columns = {
    "TRIPS", "SHOPPING SESSIONS", "RIDERS WITH SESSIONS",
    "REQUESTING SESSIONS", "REQUESTS", "ACTIVE DRIVERS"
}


def actual_analytical_type(series):
    if pd.api.types.is_datetime64_any_dtype(series):
        return "datetime"
    if pd.api.types.is_integer_dtype(series):
        return "integer"
    if pd.api.types.is_float_dtype(series):
        return "float"
    if pd.api.types.is_bool_dtype(series):
        return "boolean"
    return "string"


type_check = []
for col in df_quality.columns:
    expected = expected_types.get(col, "unknown")
    actual = actual_analytical_type(df_quality[col])
    issue = expected != "unknown" and actual != expected
    type_check.append({
        "COLUMN": col,
        "EXPECTED_TYPE": expected,
        "ACTUAL_PANDAS_DTYPE": str(df_quality[col].dtype),
        "ACTUAL_ANALYTICAL_TYPE": actual,
        "NEEDS_CONVERSION": issue,
    })

type_check = pd.DataFrame(type_check)
type_check

,COLUMN,EXPECTED_TYPE,ACTUAL_PANDAS_DTYPE,ACTUAL_ANALYTICAL_TYPE,NEEDS_CONVERSION
0,MONTH,datetime,str,string,True
1,CITY,string,str,string,False
2,TRIPS,integer,str,string,True
3,GROSS BOOKINGS (GB),float,str,string,True
4,SHOPPING SESSIONS,integer,str,string,True
5,RIDERS WITH SESSIONS,integer,str,string,True
6,REQUESTING SESSIONS,integer,str,string,True
7,REQUESTS,integer,str,string,True
8,ACTIVE DRIVERS,integer,str,string,True
9,SUPPLY HOURS,float,str,string,True


## 15. Preserve the Original CSV and Build a Typed Spreadsheet Copy

A CSV stores text values and delimiters, but it does **not** store spreadsheet cell types or number formats. Therefore, this notebook creates two separate outputs:

1. `dataset_original.csv`: an exact byte-for-byte copy of the source CSV.
2. `dataset_expected_types.xlsx`: a spreadsheet with native dates/numbers/percentages and display formats chosen so the visible cell content remains equivalent to the original CSV.

Examples:
- Raw `3,005,706` becomes numeric `3005706`, formatted as `#,##0`, so Excel still displays `3,005,706`.
- Raw `78.20%` becomes numeric `0.782`, formatted as `0.00%`, so Excel still displays `78.20%`.
- Raw `2022-01` becomes the date `2022-01-01`, formatted as `yyyy-mm`, so Excel still displays `2022-01`.

The underlying typed value must change for Excel to recognize it as a true number, percentage, or date; the displayed content is preserved through formatting.


In [16]:
from urllib.request import urlopen
from pathlib import Path

# ------------------------------------------------------------
# 1. Preserve the source CSV exactly as downloaded
# ------------------------------------------------------------
original_csv_path = Path("dataset_original.csv")

with urlopen(dataset_url) as response:
    original_csv_bytes = response.read()

original_csv_path.write_bytes(original_csv_bytes)

# ------------------------------------------------------------
# 2. Build a typed analytical copy
# ------------------------------------------------------------
df_typed = df_quality.copy()
column_actions = {}
conversion_failures = {}

for col in df_typed.columns:
    expected = expected_types.get(col)
    original_non_null = int(df_typed[col].notna().sum())

    if expected == "datetime":
        # A real spreadsheet date requires a complete date value.
        # YYYY-MM is interpreted as the first day of that month, while
        # the Excel display format later shows only YYYY-MM.
        converted = pd.to_datetime(
            df_typed[col].astype("string").str.strip(),
            format="%Y-%m",
            errors="coerce"
        )
        df_typed[col] = converted
        column_actions[col] = "Converted YYYY-MM text to native date; displayed as YYYY-MM"

    elif expected in {"integer", "float"}:
        raw = df_typed[col]

        if pd.api.types.is_numeric_dtype(raw):
            converted = pd.to_numeric(raw, errors="coerce")
            action = "Already numeric"
        else:
            cleaned_text = (
                raw.astype("string")
                   .str.strip()
                   .str.replace(",", "", regex=False)
                   .str.replace("$", "", regex=False)
            )

            if col in percentage_columns:
                had_percent_sign = cleaned_text.str.contains("%", na=False)
                cleaned_text = cleaned_text.str.replace("%", "", regex=False)
                converted = pd.to_numeric(cleaned_text, errors="coerce")

                # Excel stores percentages as decimal ratios.
                converted = converted.where(~had_percent_sign, converted / 100)
                action = "Converted percentage text to numeric ratio; displayed as percentage"
            else:
                converted = pd.to_numeric(cleaned_text, errors="coerce")
                action = "Converted numeric text to native number; display preserved with formatting"

        if expected == "integer":
            df_typed[col] = converted.round().astype("Int64")
        else:
            df_typed[col] = converted.astype("Float64")

        column_actions[col] = action

    elif expected == "string":
        df_typed[col] = df_typed[col].astype("string")
        column_actions[col] = "Stored as text"

    else:
        column_actions[col] = "No expected type configured"

    after_non_null = int(df_typed[col].notna().sum())
    conversion_failures[col] = max(original_non_null - after_non_null, 0)

typed_dtype_report = pd.DataFrame({
    "COLUMN": df_typed.columns,
    "RAW_DTYPE": [str(df_quality[c].dtype) for c in df_typed.columns],
    "EXPECTED_TYPE": [expected_types.get(c, "unknown") for c in df_typed.columns],
    "TYPED_DTYPE": [str(df_typed[c].dtype) for c in df_typed.columns],
    "ACTION": [column_actions[c] for c in df_typed.columns],
    "CONVERSION_FAILURES": [conversion_failures[c] for c in df_typed.columns],
})

print("Exact original CSV saved as:", original_csv_path)
typed_dtype_report


Exact original CSV saved as: dataset_original.csv


,COLUMN,RAW_DTYPE,EXPECTED_TYPE,TYPED_DTYPE,ACTION,CONVERSION_FAILURES
0,MONTH,str,datetime,datetime64[us],Converted YYYY-MM text to native date; display...,0
1,CITY,str,string,string,Stored as text,0
2,TRIPS,str,integer,Int64,Converted numeric text to native number; displ...,0
3,GROSS BOOKINGS (GB),str,float,Float64,Converted numeric text to native number; displ...,0
4,SHOPPING SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,0
5,RIDERS WITH SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,0
6,REQUESTING SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,0
7,REQUESTS,str,integer,Int64,Converted numeric text to native number; displ...,0
8,ACTIVE DRIVERS,str,integer,Int64,Converted numeric text to native number; displ...,0
9,SUPPLY HOURS,str,float,Float64,Converted numeric text to native number; displ...,0


## 16. Save the Typed Spreadsheet

`dataset_expected_types.xlsx` stores native spreadsheet values and applies number formats so the visible values match the original CSV as closely as possible.

A second typed CSV is intentionally **not** created because CSV files cannot retain cell types or number-format metadata.


In [17]:
import openpyxl
from openpyxl.utils.dataframe import dataframe_to_rows
from openpyxl.utils import get_column_letter

typed_xlsx_path = "dataset_expected_types.xlsx"

wb = openpyxl.Workbook()
ws = wb.active
ws.title = "dataset"

# Write headers and typed values.
for row in dataframe_to_rows(df_typed, index=False, header=True):
    ws.append(row)

header_positions = {
    cell.value: cell.column
    for cell in ws[1]
}

# Explicit display formats preserve the original visual representation.
excel_number_formats = {
    "MONTH": "yyyy-mm",
    "TRIPS": "#,##0",
    "GROSS BOOKINGS (GB)": "#,##0.00",
    "SHOPPING SESSIONS": "#,##0",
    "RIDERS WITH SESSIONS": "#,##0",
    "REQUESTING SESSIONS": "#,##0",
    "REQUESTS": "#,##0",
    "ACTIVE DRIVERS": "#,##0",
    "SUPPLY HOURS": "#,##0.00",
    "C/R": "0.00%",
    "C/RS": "0.00%",
    "RS/SS": "0.00%",
    "C/SS": "0.00%",
    "ACCEPTANCE RATE": "0.00%",
    "DRIVER CANCELATION": "0.00%",
    "SUPPLY UTILIZATION": "0.00%",
    "NPI": "0.00",
    "DRIVER PAYMENTS (AS % OF GB)": "0.00%",
    "RIDER INCENTIVES AND PROMOTIONS (AS % OF GB)": "0.00%",
    "DRIVER INCENTIVES AND PROMOTIONS (AS % OF GB)": "0.00%",
    "TAKE RATE (AS % OF GB)": "0.00%",
}

for col_name, number_format in excel_number_formats.items():
    if col_name not in header_positions:
        continue

    col_idx = header_positions[col_name]

    for row_idx in range(2, ws.max_row + 1):
        ws.cell(row=row_idx, column=col_idx).number_format = number_format

# Basic readability.
ws.freeze_panes = "A2"
ws.auto_filter.ref = ws.dimensions

for col_idx, col_name in enumerate(df_typed.columns, start=1):
    max_length = max(
        len(str(col_name)),
        max(
            (len(str(ws.cell(row=r, column=col_idx).value)) for r in range(2, ws.max_row + 1)),
            default=0
        )
    )
    ws.column_dimensions[get_column_letter(col_idx)].width = min(max(max_length + 2, 12), 34)

wb.save(typed_xlsx_path)

print("Saved exact original CSV:", original_csv_path)
print("Saved typed spreadsheet:", typed_xlsx_path)
print("Rows:", len(df_typed), "| Columns:", len(df_typed.columns))


Saved exact original CSV: dataset_original.csv
Saved typed spreadsheet: dataset_expected_types.xlsx
Rows: 168 | Columns: 22


## 17. Original vs. Typed Copy Comparison

This table shows the raw pandas dtype, the expected type, the native dtype used in the typed copy, and whether any non-null source values failed conversion.


In [18]:
before_after = []

for col in df_quality.columns:
    raw_non_null = df_quality[col].dropna()
    typed_non_null = df_typed[col].dropna()

    before_after.append({
        "COLUMN": col,
        "RAW_DTYPE": str(df_quality[col].dtype),
        "EXPECTED_TYPE": expected_types.get(col, "unknown"),
        "TYPED_DTYPE": str(df_typed[col].dtype),
        "ACTION": column_actions[col],
        "RAW_EXAMPLE": raw_non_null.iloc[0] if not raw_non_null.empty else None,
        "TYPED_VALUE_EXAMPLE": typed_non_null.iloc[0] if not typed_non_null.empty else None,
        "EXCEL_DISPLAY_FORMAT": excel_number_formats.get(col, "General"),
        "CONVERSION_FAILURES": conversion_failures[col],
    })

before_after = pd.DataFrame(before_after)
before_after


,COLUMN,RAW_DTYPE,EXPECTED_TYPE,TYPED_DTYPE,ACTION,RAW_EXAMPLE,TYPED_VALUE_EXAMPLE,EXCEL_DISPLAY_FORMAT,CONVERSION_FAILURES
0,MONTH,str,datetime,datetime64[us],Converted YYYY-MM text to native date; display...,2022-01,2022-01-01 00:00:00,yyyy-mm,0
1,CITY,str,string,string,Stored as text,Vila Belmiro,Vila Belmiro,General,0
2,TRIPS,str,integer,Int64,Converted numeric text to native number; displ...,"3,005,706",3005706,"#,##0",0
3,GROSS BOOKINGS (GB),str,float,Float64,Converted numeric text to native number; displ...,"56,529,569",56529569.0,"#,##0.00",0
4,SHOPPING SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,"5,283,006",5283006,"#,##0",0
5,RIDERS WITH SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,"711,432",711432,"#,##0",0
6,REQUESTING SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,"3,257,826",3257826,"#,##0",0
7,REQUESTS,str,integer,Int64,Converted numeric text to native number; displ...,"3,844,423",3844423,"#,##0",0
8,ACTIVE DRIVERS,str,integer,Int64,Converted numeric text to native number; displ...,"18,795",18795,"#,##0",0
9,SUPPLY HOURS,str,float,Float64,Converted numeric text to native number; displ...,"1,564,816",1564816.0,"#,##0.00",0


## Data Type Check

Expected data types are defined manually and compared against the current pandas dtype.


In [19]:
# ============================================================
# 18. DATA TYPE CHECK
# METRICS | CURRENT DATA | EXPECTED DATA | CORRECT
# ============================================================

from pathlib import Path
import html
import pandas as pd


# ============================================================
# 1. Expected data types defined manually
# ============================================================

expected_types = {
    "MONTH": "date",
    "CITY": "string",

    "TRIPS": "int",
    "GROSS BOOKINGS (GB)": "int",
    "SHOPPING SESSIONS": "int",
    "RIDERS WITH SESSIONS": "int",
    "REQUESTING SESSIONS": "int",
    "REQUESTS": "int",
    "ACTIVE DRIVERS": "int",
    "SUPPLY HOURS": "int",

    "C/R": "float",
    "C/RS": "float",
    "RS/SS": "float",
    "C/SS": "float",
    "ACCEPTANCE RATE": "float",
    "DRIVER CANCELATION": "float",
    "SUPPLY UTILIZATION": "float",

    "NPI": "float64",

    "DRIVER PAYMENTS (AS % OF GB)": "float",
    "RIDER INCENTIVES AND PROMOTIONS (AS % OF GB)": "float",
    "DRIVER INCENTIVES AND PROMOTIONS (AS % OF GB)": "float",
    "TAKE RATE (AS % OF GB)": "float",
}


# ============================================================
# 2. Build the four-column validation table
# ============================================================

data_type_table = pd.DataFrame({
    "METRICS": df_quality.columns,
    "CURRENT DATA": [
        str(df_quality[column].dtype)
        for column in df_quality.columns
    ],
    "EXPECTED DATA": [
        expected_types.get(column, "")
        for column in df_quality.columns
    ],
})

data_type_table["CORRECT"] = (
    data_type_table["CURRENT DATA"]
    == data_type_table["EXPECTED DATA"]
).map({
    True: "Yes",
    False: "No",
})


# ============================================================
# 3. Build HTML rows
# ============================================================

data_type_rows_html = ""

for _, row in data_type_table.iterrows():

    correct_class = (
        "dtype-correct"
        if row["CORRECT"] == "Yes"
        else "dtype-incorrect"
    )

    data_type_rows_html += f"""
    <tr>
        <td class="dtype-metric">
            {html.escape(str(row["METRICS"]))}
        </td>

        <td>
            {html.escape(str(row["CURRENT DATA"]))}
        </td>

        <td>
            {html.escape(str(row["EXPECTED DATA"]))}
        </td>

        <td>
            <span class="dtype-badge {correct_class}">
                {html.escape(str(row["CORRECT"]))}
            </span>
        </td>
    </tr>
    """


# ============================================================
# 4. Build Data Type Check HTML section
# ============================================================

data_type_section_html = f"""
<!-- DATA_TYPE_CHECK_START -->

<style>
    .dtype-section {{
        max-width: 1400px;
        margin: 28px auto;
        background: #ffffff;
        border-radius: 16px;
        box-shadow:
            0 4px 18px rgba(0, 0, 0, 0.07);
        padding: 30px;
        color: #111827;
        font-family:
            Arial,
            Helvetica,
            sans-serif;
    }}

    .dtype-section h2 {{
        text-align: center;
        margin: 0 0 10px;
        font-size: 25px;
    }}

    .dtype-description {{
        max-width: 900px;
        margin: 0 auto 24px;
        text-align: center;
        color: #6b7280;
        font-size: 14px;
        line-height: 1.5;
    }}

    .dtype-table-wrapper {{
        overflow-x: auto;
        border: 1px solid #e5e7eb;
        border-radius: 12px;
    }}

    .dtype-table {{
        width: 100%;
        border-collapse: collapse;
        min-width: 760px;
    }}

    .dtype-table th {{
        background: #f9fafb;
        color: #6b7280;
        text-transform: uppercase;
        font-size: 11px;
        letter-spacing: 0.45px;
        padding: 14px 16px;
        text-align: left;
        border-bottom: 1px solid #e5e7eb;
    }}

    .dtype-table td {{
        padding: 14px 16px;
        border-bottom: 1px solid #f0f1f3;
        font-size: 13px;
    }}

    .dtype-metric {{
        font-weight: 700;
    }}

    .dtype-badge {{
        display: inline-block;
        padding: 5px 11px;
        border-radius: 999px;
        font-size: 11px;
        font-weight: 700;
    }}

    .dtype-correct {{
        background: #dcfae6;
        color: #067647;
    }}

    .dtype-incorrect {{
        background: #fee4e2;
        color: #b42318;
    }}
</style>


<section class="dtype-section">

    <h2>
        DATA TYPE CHECK
    </h2>

    <p class="dtype-description">
        Compares how pandas currently reads each dataset column
        against the expected data type defined manually for the analysis.
    </p>

    <div class="dtype-table-wrapper">

        <table class="dtype-table">

            <thead>
                <tr>
                    <th>Metrics</th>
                    <th>Current Data</th>
                    <th>Expected Data</th>
                    <th>Correct</th>
                </tr>
            </thead>

            <tbody>
                {data_type_rows_html}
            </tbody>

        </table>

    </div>

</section>

<!-- DATA_TYPE_CHECK_END -->
"""


# ============================================================
# 5. Insert the table into data_quality.html
# Section 20 is preserved
# ============================================================

data_quality_path = Path("data_quality.html")

if not data_quality_path.exists():
    raise FileNotFoundError(
        "data_quality.html was not found. "
        "Run the previous Data Quality cells first."
    )

current_html = data_quality_path.read_text(
    encoding="utf-8"
)

start_marker = "<!-- DATA_TYPE_CHECK_START -->"
end_marker = "<!-- DATA_TYPE_CHECK_END -->"

# Remove a prior generated copy of this four-column section.
if start_marker in current_html and end_marker in current_html:
    before_section = current_html.split(
        start_marker,
        1,
    )[0]

    after_section = current_html.split(
        end_marker,
        1,
    )[1]

    current_html = before_section + after_section


# Remove the old Data Type Check card that contained
# RAW DTYPE / TYPED DTYPE / CONVERTED, if present.
old_title_candidates = [
    "<h1>DATA TYPE CHECK</h1>",
    "<h2>DATA TYPE CHECK</h2>",
]

for old_title in old_title_candidates:

    if old_title not in current_html:
        continue

    title_position = current_html.index(old_title)

    old_start = current_html.rfind(
        "<section",
        0,
        title_position,
    )

    old_end = current_html.find(
        "</section>",
        title_position,
    )

    if old_start != -1 and old_end != -1:
        current_html = (
            current_html[:old_start]
            + current_html[
                old_end + len("</section>"):
            ]
        )

        break


# Insert this table immediately before Section 20.
ratio_marker = "<!-- FUNNEL_RATIO_VALIDATION_START -->"

if ratio_marker in current_html:

    updated_html = current_html.replace(
        ratio_marker,
        data_type_section_html
        + "\n"
        + ratio_marker,
        1,
    )

elif "</body>" in current_html:

    updated_html = current_html.replace(
        "</body>",
        data_type_section_html
        + "\n</body>",
        1,
    )

else:

    updated_html = (
        current_html
        + data_type_section_html
    )


data_quality_path.write_text(
    updated_html,
    encoding="utf-8",
)


print(
    f"Updated: {data_quality_path.resolve()}"
)

print(
    "Data Type Check restored:"
)

print(
    "METRICS | CURRENT DATA | EXPECTED DATA | CORRECT"
)

print(
    "Section 20 was preserved."
)


data_type_table


Updated: /Users/elvamariaramosgonzalez/Desktop/dashboard-demo/data_quality.html
Data Type Check restored:
METRICS | CURRENT DATA | EXPECTED DATA | CORRECT
Section 20 was preserved.


,METRICS,CURRENT DATA,EXPECTED DATA,CORRECT
0,MONTH,str,date,No
1,CITY,str,string,No
2,TRIPS,str,int,No
3,GROSS BOOKINGS (GB),str,int,No
4,SHOPPING SESSIONS,str,int,No
5,RIDERS WITH SESSIONS,str,int,No
6,REQUESTING SESSIONS,str,int,No
7,REQUESTS,str,int,No
8,ACTIVE DRIVERS,str,int,No
9,SUPPLY HOURS,str,int,No


## 19. Final dtype validation

This final check confirms the pandas dtype of every cleaned column and surfaces any conversion failures.

In [20]:
final_validation = typed_dtype_report.copy()
final_validation["HAS_CONVERSION_FAILURES"] = final_validation["CONVERSION_FAILURES"] > 0
final_validation

,COLUMN,RAW_DTYPE,EXPECTED_TYPE,TYPED_DTYPE,ACTION,CONVERSION_FAILURES,HAS_CONVERSION_FAILURES
0,MONTH,str,datetime,datetime64[us],Converted YYYY-MM text to native date; display...,0,False
1,CITY,str,string,string,Stored as text,0,False
2,TRIPS,str,integer,Int64,Converted numeric text to native number; displ...,0,False
3,GROSS BOOKINGS (GB),str,float,Float64,Converted numeric text to native number; displ...,0,False
4,SHOPPING SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,0,False
5,RIDERS WITH SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,0,False
6,REQUESTING SESSIONS,str,integer,Int64,Converted numeric text to native number; displ...,0,False
7,REQUESTS,str,integer,Int64,Converted numeric text to native number; displ...,0,False
8,ACTIVE DRIVERS,str,integer,Int64,Converted numeric text to native number; displ...,0,False
9,SUPPLY HOURS,str,float,Float64,Converted numeric text to native number; displ...,0,False


## 20. Funnel Ratio Validation — Row-Level Detail

For every `MONTH + CITY + METRIC`, this section recalculates the reported funnel ratio, calculates the absolute difference, and marks `CORRECT` as `True` when the difference is less than or equal to `0.001`. It then updates `data_quality.html`.


In [21]:
# ============================================================
# FUNNEL RATIO VALIDATION — ROW-LEVEL DETAIL
# ============================================================

from pathlib import Path
import html
import json
import numpy as np
import pandas as pd

ratio_tolerance = 0.001

ratio_definitions = {
    "C/R": {
        "numerator": "TRIPS",
        "denominator": "REQUESTS",
        "formula": "TRIPS / REQUESTS",
    },
    "C/RS": {
        "numerator": "TRIPS",
        "denominator": "REQUESTING SESSIONS",
        "formula": "TRIPS / REQUESTING SESSIONS",
    },
    "RS/SS": {
        "numerator": "REQUESTING SESSIONS",
        "denominator": "SHOPPING SESSIONS",
        "formula": "REQUESTING SESSIONS / SHOPPING SESSIONS",
    },
    "C/SS": {
        "numerator": "TRIPS",
        "denominator": "SHOPPING SESSIONS",
        "formula": "TRIPS / SHOPPING SESSIONS",
    },
}

required_ratio_columns = {
    "MONTH",
    "CITY",
    "TRIPS",
    "REQUESTS",
    "REQUESTING SESSIONS",
    "SHOPPING SESSIONS",
    "C/R",
    "C/RS",
    "RS/SS",
    "C/SS",
}

missing_ratio_columns = sorted(
    required_ratio_columns - set(df_quality.columns)
)

if missing_ratio_columns:
    raise KeyError(
        f"Missing columns required for ratio validation: {missing_ratio_columns}"
    )


def parse_numeric_for_validation(series):
    """Parse numeric values without modifying df_quality."""
    if pd.api.types.is_numeric_dtype(series):
        return pd.to_numeric(series, errors="coerce")

    cleaned = (
        series.astype("string")
        .str.strip()
        .str.replace(",", "", regex=False)
        .str.replace("$", "", regex=False)
    )

    return pd.to_numeric(cleaned, errors="coerce")


def parse_ratio_for_validation(series):
    """Parse ratios stored as decimals or percentages."""
    if pd.api.types.is_numeric_dtype(series):
        numeric = pd.to_numeric(series, errors="coerce")
        return numeric.where(numeric.abs() <= 1, numeric / 100)

    raw = series.astype("string").str.strip()
    contains_percent = raw.str.contains("%", na=False)

    cleaned = (
        raw.str.replace("%", "", regex=False)
        .str.replace(",", "", regex=False)
    )

    numeric = pd.to_numeric(cleaned, errors="coerce")
    numeric = numeric.where(~contains_percent, numeric / 100)
    numeric = numeric.where(numeric.abs() <= 1, numeric / 100)

    return numeric


validation_frames = []

for metric, definition in ratio_definitions.items():
    numerator = parse_numeric_for_validation(
        df_quality[definition["numerator"]]
    )

    denominator = parse_numeric_for_validation(
        df_quality[definition["denominator"]]
    )

    reported = parse_ratio_for_validation(
        df_quality[metric]
    )

    recalculated = numerator.div(
        denominator.where(denominator.ne(0))
    )

    difference = (reported - recalculated).abs()

    correct = difference.le(ratio_tolerance).astype("boolean")
    correct = correct.where(
        reported.notna() & recalculated.notna(),
        pd.NA,
    )

    validation_frames.append(
        pd.DataFrame({
            "MONTH": df_quality["MONTH"],
            "CITY": df_quality["CITY"],
            "METRIC": metric,
            "FORMULA": definition["formula"],
            "REPORTED": reported,
            "RECALCULATED": recalculated,
            "DIFFERENCE": difference,
            "CORRECT": correct,
        })
    )

ratio_validation_detail = pd.concat(
    validation_frames,
    ignore_index=True,
)

# Preserve a stable initial order.
ratio_validation_detail = ratio_validation_detail.sort_values(
    ["MONTH", "CITY", "METRIC"],
    kind="stable",
).reset_index(drop=True)

# Summary metrics.
testable_mask = ratio_validation_detail["CORRECT"].notna()
total_validations = int(testable_mask.sum())
correct_validations = int(
    ratio_validation_detail.loc[testable_mask, "CORRECT"].eq(True).sum()
)
incorrect_validations = int(
    ratio_validation_detail.loc[testable_mask, "CORRECT"].eq(False).sum()
)
validation_accuracy = (
    correct_validations / total_validations
    if total_validations > 0
    else np.nan
)

# Filter options.
month_options = sorted(
    ratio_validation_detail["MONTH"]
    .dropna()
    .astype(str)
    .unique()
    .tolist()
)

city_options = sorted(
    ratio_validation_detail["CITY"]
    .dropna()
    .astype(str)
    .unique()
    .tolist()
)

metric_options = list(ratio_definitions.keys())


def build_select_options(values):
    return "".join(
        f'<option value="{html.escape(value)}">{html.escape(value)}</option>'
        for value in values
    )


month_options_html = build_select_options(month_options)
city_options_html = build_select_options(city_options)
metric_options_html = build_select_options(metric_options)

# Build detailed HTML rows.
ratio_rows_html = ""

for _, row in ratio_validation_detail.iterrows():
    if pd.isna(row["CORRECT"]):
        correct_label = "N/A"
        correct_class = "ratio-na"
    elif bool(row["CORRECT"]):
        correct_label = "True"
        correct_class = "ratio-true"
    else:
        correct_label = "False"
        correct_class = "ratio-false"

    reported_label = (
        f'{row["REPORTED"]:.4f}'
        if pd.notna(row["REPORTED"])
        else "N/A"
    )

    recalculated_label = (
        f'{row["RECALCULATED"]:.4f}'
        if pd.notna(row["RECALCULATED"])
        else "N/A"
    )

    difference_label = (
        f'{row["DIFFERENCE"]:.4f}'
        if pd.notna(row["DIFFERENCE"])
        else "N/A"
    )

    month_value = html.escape(str(row["MONTH"]))
    city_value = html.escape(str(row["CITY"]))
    metric_value = html.escape(str(row["METRIC"]))

    ratio_rows_html += f"""
    <tr
        data-month="{month_value}"
        data-city="{city_value}"
        data-metric="{metric_value}"
        data-correct="{correct_label}"
    >
        <td data-sort-value="{month_value}">{month_value}</td>
        <td data-sort-value="{city_value}">{city_value}</td>
        <td class="ratio-metric" data-sort-value="{metric_value}">
            {metric_value}
        </td>
        <td>{html.escape(str(row['FORMULA']))}</td>
        <td data-sort-value="{reported_label}">{reported_label}</td>
        <td data-sort-value="{recalculated_label}">{recalculated_label}</td>
        <td data-sort-value="{difference_label}">{difference_label}</td>
        <td data-sort-value="{correct_label}">
            <span class="ratio-badge {correct_class}">{correct_label}</span>
        </td>
    </tr>
    """

accuracy_label = (
    f"{validation_accuracy:.1%}"
    if pd.notna(validation_accuracy)
    else "N/A"
)

ratio_section_html = f"""
<!-- FUNNEL_RATIO_VALIDATION_START -->
<style>
    .ratio-section {{
        max-width: 1400px;
        margin: 28px auto;
        background: #ffffff;
        border-radius: 16px;
        box-shadow: 0 4px 18px rgba(0, 0, 0, 0.07);
        padding: 30px;
        color: #111827;
        font-family: Arial, Helvetica, sans-serif;
    }}

    .ratio-section h2 {{
        text-align: center;
        margin: 0 0 10px;
        font-size: 25px;
    }}

    .ratio-description {{
        max-width: 980px;
        margin: 0 auto 24px;
        text-align: center;
        color: #6b7280;
        font-size: 14px;
        line-height: 1.5;
    }}

    .ratio-disclaimer {{
        max-width: 980px;
        margin: 0 auto 24px;
        padding: 16px 18px;
        border: 1px solid #fdb022;
        border-left: 5px solid #f79009;
        border-radius: 10px;
        background: #fffaeb;
        color: #7a2e0e;
        font-size: 13px;
        line-height: 1.55;
    }}

    .ratio-disclaimer strong {{
        color: #b54708;
    }}

    .ratio-summary-grid {{
        display: grid;
        grid-template-columns: repeat(4, minmax(0, 1fr));
        gap: 16px;
        margin-bottom: 24px;
    }}

    .ratio-summary-card {{
        border: 1px solid #e5e7eb;
        border-radius: 12px;
        padding: 18px;
        text-align: center;
    }}

    .ratio-summary-label {{
        color: #6b7280;
        font-size: 11px;
        text-transform: uppercase;
        letter-spacing: 0.45px;
        margin-bottom: 8px;
    }}

    .ratio-summary-value {{
        font-size: 24px;
        font-weight: 700;
    }}

    .ratio-filter-grid {{
        display: grid;
        grid-template-columns: repeat(5, minmax(0, 1fr));
        gap: 14px;
        margin-bottom: 16px;
        align-items: end;
    }}

    .ratio-filter-group label {{
        display: block;
        margin-bottom: 6px;
        color: #475467;
        font-size: 11px;
        font-weight: 700;
        text-transform: uppercase;
        letter-spacing: 0.35px;
    }}

    .ratio-filter-group select,
    .ratio-filter-group input {{
        width: 100%;
        height: 40px;
        border: 1px solid #d0d5dd;
        border-radius: 8px;
        padding: 0 11px;
        background: #ffffff;
        color: #101828;
        font-size: 13px;
    }}

    .ratio-reset-button {{
        width: 100%;
        height: 40px;
        border: 1px solid #d0d5dd;
        border-radius: 8px;
        background: #ffffff;
        color: #344054;
        font-size: 13px;
        font-weight: 700;
        cursor: pointer;
    }}

    .ratio-reset-button:hover {{
        background: #f9fafb;
    }}

    .ratio-visible-count {{
        margin: 0 0 10px;
        color: #667085;
        font-size: 12px;
        text-align: right;
    }}

    .ratio-table-wrapper {{
        overflow: auto;
        max-height: 720px;
        border: 1px solid #e5e7eb;
        border-radius: 12px;
    }}

    .ratio-table {{
        width: 100%;
        border-collapse: collapse;
        min-width: 950px;
    }}

    .ratio-table th {{
        position: sticky;
        top: 0;
        background: #f9fafb;
        color: #6b7280;
        text-transform: uppercase;
        font-size: 11px;
        letter-spacing: 0.45px;
        padding: 14px 16px;
        text-align: left;
        border-bottom: 1px solid #e5e7eb;
        z-index: 1;
        user-select: none;
    }}

    .ratio-table th[data-sort-index] {{
        cursor: pointer;
    }}

    .ratio-table th[data-sort-index]:hover {{
        background: #f2f4f7;
        color: #344054;
    }}

    .ratio-sort-indicator {{
        margin-left: 5px;
        font-size: 10px;
    }}

    .ratio-table td {{
        padding: 13px 16px;
        border-bottom: 1px solid #f0f1f3;
        font-size: 13px;
        white-space: nowrap;
    }}

    .ratio-metric {{
        font-weight: 700;
    }}

    .ratio-badge {{
        display: inline-block;
        padding: 5px 11px;
        border-radius: 999px;
        font-size: 11px;
        font-weight: 700;
    }}

    .ratio-true {{
        background: #dcfae6;
        color: #067647;
    }}

    .ratio-false {{
        background: #fee4e2;
        color: #b42318;
    }}

    .ratio-na {{
        background: #f2f4f7;
        color: #475467;
    }}

    @media (max-width: 1050px) {{
        .ratio-filter-grid {{
            grid-template-columns: repeat(2, minmax(0, 1fr));
        }}
    }}

    @media (max-width: 900px) {{
        .ratio-summary-grid {{
            grid-template-columns: repeat(2, minmax(0, 1fr));
        }}
    }}

    @media (max-width: 620px) {{
        .ratio-filter-grid {{
            grid-template-columns: 1fr;
        }}
    }}
</style>

<section class="ratio-section">
    <h2>FUNNEL RATIO VALIDATION</h2>

    <p class="ratio-description">
        For every dataset row, the dashboard recalculates C/R, C/RS,
        RS/SS and C/SS, compares each result against the reported value,
        calculates the absolute difference, and marks CORRECT as True
        when the difference is less than or equal to {ratio_tolerance:.3f}.
        Use the filters below or click a column header to sort the table.
    </p>

    <div class="ratio-disclaimer">
        <strong>Validation tolerance: 0.001.</strong>
        This threshold is applied because the reported compound metrics
        may include rounding during calculation or storage. Minor
        differences between reported and recalculated values are therefore
        treated as acceptable rather than as data-quality failures.
    </div>

    <div class="ratio-summary-grid">
        <div class="ratio-summary-card">
            <div class="ratio-summary-label">Total Validations</div>
            <div class="ratio-summary-value">{total_validations:,}</div>
        </div>

        <div class="ratio-summary-card">
            <div class="ratio-summary-label">Correct</div>
            <div class="ratio-summary-value">{correct_validations:,}</div>
        </div>

        <div class="ratio-summary-card">
            <div class="ratio-summary-label">Incorrect</div>
            <div class="ratio-summary-value">{incorrect_validations:,}</div>
        </div>

        <div class="ratio-summary-card">
            <div class="ratio-summary-label">Accuracy</div>
            <div class="ratio-summary-value">{accuracy_label}</div>
        </div>
    </div>

    <div class="ratio-filter-grid">
        <div class="ratio-filter-group">
            <label for="ratio-month-filter">Month</label>
            <select id="ratio-month-filter">
                <option value="">All months</option>
                {month_options_html}
            </select>
        </div>

        <div class="ratio-filter-group">
            <label for="ratio-city-filter">City</label>
            <select id="ratio-city-filter">
                <option value="">All cities</option>
                {city_options_html}
            </select>
        </div>

        <div class="ratio-filter-group">
            <label for="ratio-metric-filter">Metric</label>
            <select id="ratio-metric-filter">
                <option value="">All metrics</option>
                {metric_options_html}
            </select>
        </div>

        <div class="ratio-filter-group">
            <label for="ratio-correct-filter">Correct</label>
            <select id="ratio-correct-filter">
                <option value="">All results</option>
                <option value="True">True</option>
                <option value="False">False</option>
                <option value="N/A">N/A</option>
            </select>
        </div>

        <div class="ratio-filter-group">
            <button
                type="button"
                id="ratio-reset-filters"
                class="ratio-reset-button"
            >
                Reset filters
            </button>
        </div>
    </div>

    <div id="ratio-visible-count" class="ratio-visible-count"></div>

    <div class="ratio-table-wrapper">
        <table id="ratio-validation-table" class="ratio-table">
            <thead>
                <tr>
                    <th data-sort-index="0" data-sort-type="text">
                        Month<span class="ratio-sort-indicator"></span>
                    </th>
                    <th data-sort-index="1" data-sort-type="text">
                        City<span class="ratio-sort-indicator"></span>
                    </th>
                    <th data-sort-index="2" data-sort-type="text">
                        Metric<span class="ratio-sort-indicator"></span>
                    </th>
                    <th>Formula</th>
                    <th data-sort-index="4" data-sort-type="number">
                        Reported<span class="ratio-sort-indicator"></span>
                    </th>
                    <th data-sort-index="5" data-sort-type="number">
                        Recalculated<span class="ratio-sort-indicator"></span>
                    </th>
                    <th data-sort-index="6" data-sort-type="number">
                        Difference<span class="ratio-sort-indicator"></span>
                    </th>
                    <th data-sort-index="7" data-sort-type="text">
                        Correct<span class="ratio-sort-indicator"></span>
                    </th>
                </tr>
            </thead>
            <tbody>
                {ratio_rows_html}
            </tbody>
        </table>
    </div>
</section>

<script>
(function () {{
    const table = document.getElementById("ratio-validation-table");

    if (!table) {{
        return;
    }}

    const tbody = table.querySelector("tbody");
    const monthFilter = document.getElementById("ratio-month-filter");
    const cityFilter = document.getElementById("ratio-city-filter");
    const metricFilter = document.getElementById("ratio-metric-filter");
    const correctFilter = document.getElementById("ratio-correct-filter");
    const resetButton = document.getElementById("ratio-reset-filters");
    const visibleCount = document.getElementById("ratio-visible-count");

    let sortIndex = null;
    let sortDirection = "asc";

    function applyFilters() {{
        const monthValue = monthFilter.value;
        const cityValue = cityFilter.value;
        const metricValue = metricFilter.value;
        const correctValue = correctFilter.value;

        let visibleRows = 0;

        Array.from(tbody.rows).forEach((row) => {{
            const matchesMonth =
                !monthValue || row.dataset.month === monthValue;

            const matchesCity =
                !cityValue || row.dataset.city === cityValue;

            const matchesMetric =
                !metricValue || row.dataset.metric === metricValue;

            const matchesCorrect =
                !correctValue || row.dataset.correct === correctValue;

            const visible =
                matchesMonth
                && matchesCity
                && matchesMetric
                && matchesCorrect;

            row.style.display = visible ? "" : "none";

            if (visible) {{
                visibleRows += 1;
            }}
        }});

        visibleCount.textContent =
            `Showing ${{visibleRows.toLocaleString()}} of `
            + `${{tbody.rows.length.toLocaleString()}} validations`;
    }}

    function getCellValue(row, index, type) {{
        const cell = row.cells[index];
        const rawValue =
            cell.dataset.sortValue || cell.textContent.trim();

        if (type === "number") {{
            const numericValue = Number(rawValue);
            return Number.isFinite(numericValue)
                ? numericValue
                : Number.POSITIVE_INFINITY;
        }}

        return rawValue.toLowerCase();
    }}

    function sortTable(header) {{
        const nextSortIndex = Number(header.dataset.sortIndex);
        const sortType = header.dataset.sortType || "text";

        if (sortIndex === nextSortIndex) {{
            sortDirection =
                sortDirection === "asc" ? "desc" : "asc";
        }} else {{
            sortIndex = nextSortIndex;
            sortDirection = "asc";
        }}

        const rows = Array.from(tbody.rows);

        rows.sort((rowA, rowB) => {{
            const valueA = getCellValue(
                rowA,
                sortIndex,
                sortType,
            );

            const valueB = getCellValue(
                rowB,
                sortIndex,
                sortType,
            );

            if (valueA < valueB) {{
                return sortDirection === "asc" ? -1 : 1;
            }}

            if (valueA > valueB) {{
                return sortDirection === "asc" ? 1 : -1;
            }}

            return 0;
        }});

        rows.forEach((row) => tbody.appendChild(row));

        table.querySelectorAll(".ratio-sort-indicator")
            .forEach((indicator) => {{
                indicator.textContent = "";
            }});

        const indicator = header.querySelector(
            ".ratio-sort-indicator"
        );

        if (indicator) {{
            indicator.textContent =
                sortDirection === "asc" ? "▲" : "▼";
        }}
    }}

    [
        monthFilter,
        cityFilter,
        metricFilter,
        correctFilter,
    ].forEach((filter) => {{
        filter.addEventListener("change", applyFilters);
    }});

    resetButton.addEventListener("click", () => {{
        monthFilter.value = "";
        cityFilter.value = "";
        metricFilter.value = "";
        correctFilter.value = "";
        applyFilters();
    }});

    table.querySelectorAll("th[data-sort-index]")
        .forEach((header) => {{
            header.addEventListener("click", () => {{
                sortTable(header);
            }});
        }});

    applyFilters();
}})();
</script>
<!-- FUNNEL_RATIO_VALIDATION_END -->
"""

# Update the HTML generated by the prior notebook sections.
data_quality_path = Path("data_quality.html")

if not data_quality_path.exists():
    raise FileNotFoundError(
        "data_quality.html was not found. Run the prior Data Quality cells first."
    )

current_html = data_quality_path.read_text(encoding="utf-8")

start_marker = "<!-- FUNNEL_RATIO_VALIDATION_START -->"
end_marker = "<!-- FUNNEL_RATIO_VALIDATION_END -->"

if start_marker in current_html and end_marker in current_html:
    before = current_html.split(start_marker, 1)[0]
    after = current_html.split(end_marker, 1)[1]
    current_html = before + after

if "</body>" in current_html:
    updated_html = current_html.replace(
        "</body>",
        ratio_section_html + "\n</body>",
        1,
    )
else:
    updated_html = current_html + ratio_section_html

data_quality_path.write_text(
    updated_html,
    encoding="utf-8",
)

print(f"Updated: {data_quality_path.resolve()}")
print(f"Tolerance: {ratio_tolerance}")
print(f"Total validations: {total_validations:,}")
print(f"Correct: {correct_validations:,}")
print(f"Incorrect: {incorrect_validations:,}")
print(f"Accuracy: {accuracy_label}")
print("Filters: MONTH, CITY, METRIC, CORRECT")
print("Sorting: click the sortable table headers")

ratio_validation_detail.head(12)


Updated: /Users/elvamariaramosgonzalez/Desktop/dashboard-demo/data_quality.html
Tolerance: 0.001
Total validations: 672
Correct: 672
Incorrect: 0
Accuracy: 100.0%
Filters: MONTH, CITY, METRIC, CORRECT
Sorting: click the sortable table headers


,MONTH,CITY,METRIC,FORMULA,REPORTED,RECALCULATED,DIFFERENCE,CORRECT
0,2022-01,Azteca,C/R,TRIPS / REQUESTS,0.696,0.695994,0.000006,True
1,2022-01,Azteca,C/RS,TRIPS / REQUESTING SESSIONS,0.87,0.86983,0.00017,True
2,2022-01,Azteca,C/SS,TRIPS / SHOPPING SESSIONS,0.493,0.492691,0.000309,True
3,2022-01,Azteca,RS/SS,REQUESTING SESSIONS / SHOPPING SESSIONS,0.566,0.566423,0.000423,True
4,2022-01,Lusail,C/R,TRIPS / REQUESTS,0.757,0.756607,0.000393,True
5,2022-01,Lusail,C/RS,TRIPS / REQUESTING SESSIONS,0.921,0.921375,0.000375,True
6,2022-01,Lusail,C/SS,TRIPS / SHOPPING SESSIONS,0.549,0.549163,0.000163,True
7,2022-01,Lusail,RS/SS,REQUESTING SESSIONS / SHOPPING SESSIONS,0.596,0.596025,0.000025,True
8,2022-01,Maracanã,C/R,TRIPS / REQUESTS,0.761,0.761401,0.000401,True
9,2022-01,Maracanã,C/RS,TRIPS / REQUESTING SESSIONS,0.909,0.909126,0.000126,True
